### PR TITLE
Fix most Doxygen warnings

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

This fixes almost all warnings in the external and internal API docs.

The only remaining issue is that Doxygen doesn't like raw string literals in our *.md files, e.g.
```
/home/eelco/Dev/nix-master/src/libstore/local-overlay-store.md:48: warning: Found unknown command '@docroot'
/home/eelco/Dev/nix-master/src/libstore/local-overlay-store.md:53: warning: Found unknown command '@docroot'
/home/eelco/Dev/nix-master/src/libstore/local-overlay-store.md:68: warning: Found unknown command '@docroot'
...
/home/eelco/Dev/nix-master/src/nix/registry-pin.md:46: warning: end of comment block while expecting command </tt>
```

We can set `EXCLUDE_PATTERNS = *.md`, but that just leads to a different kind of warning (namely that it can't find the `#include`d *.md files). This might be fixed in an upcoming Doxygen release (https://github.com/doxygen/doxygen/issues/11166).

# Context

Helps with #11841.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by Mic92** [comment]
From CI:

```
       > ../worker-protocol.cc:42: error: unterminated #ifndef
       >    42 | #ifndef DOXYGEN_SKIP
       >       |
```

**Comment by edolstra** [comment]
Oops, I fixed that but forgot to push.

File: src/libutil/file-system.cc [hunk 1]
```diff
@@ -29,7 +29,7 @@
 
 namespace nix {
 
-namespace fs { using namespace std::filesystem; }
+namespace fs = std::filesystem;
```
Comment (Ericson2314) [inline src/libutil/file-system.cc:32]: What's going on here? There are very specific reasons for doing this I don't want to change just to make Doxygen happy.

File: src/libutil/file-system.hh [hunk 2]
```diff
@@ -126,8 +126,6 @@ std::optional<struct stat> maybeLstat(const Path & path);
  */
 bool pathExists(const Path & path);
 
-namespace fs {
-
```
Comment (Ericson2314) [inline src/libutil/file-system.hh:130]: Do not want to change this

File: src/libutil/tarfile.cc [hunk 3]
```diff
@@ -8,9 +8,7 @@
 
 namespace nix {
 
-namespace fs {
-using namespace std::filesystem;
-}
+namespace fs = std::filesystem;
```
Comment (Ericson2314) [inline src/libutil/tarfile.cc:11]: Do not want to change this

File: src/libutil/tarfile.cc [hunk 4]
```diff
@@ -106,7 +104,7 @@ TarArchive::TarArchive(Source & source, bool raw, std::optional<std::string> com
         "Failed to open archive (%s)");
 }
 
-TarArchive::TarArchive(const fs::path & path)
+TarArchive::TarArchive(const std::filesystem::path & path)
```
Comment (Ericson2314) [inline src/libutil/tarfile.cc:107]: Do not want to change this

File: src/nix/eval.cc [hunk 5]
```diff
@@ -77,7 +77,7 @@ struct CmdEval : MixJSON, InstallableValueCommand, MixReadOnlyOption
         if (writeTo) {
             stopProgressBar();
 
-            if (fs::symlink_exists(*writeTo))
+            if (nix::symlink_exists(*writeTo))
```
Comment (Ericson2314) [inline src/nix/eval.cc:80]: Do not want to change this

**Review by Ericson2314 — CHANGES_REQUESTED** [review]
Please remove the `fs` stuff unless it is absolutely critical. I would like to discuss that in a separate PR instead.

File: src/libutil/tarfile.cc (hunk 4 above)
Comment (edolstra) [reply src/libutil/tarfile.cc:107]: Why not? The header file uses `std::filesystem::path` as well:
```
    explicit TarArchive(const std::filesystem::path & path);
```

Comment (Ericson2314) [reply src/libutil/tarfile.cc:107]: I do care less in the `cc` file; I just think if we're bothering to make the `fs` shorthand we might as well use it.

File: src/libutil/file-system.cc (hunk 1 above)
Comment (Ericson2314) [reply src/libutil/file-system.cc:32]: I wanted to make sure to never clone in in a header, precisely so it wouldn't show up in doxygen or downstream code.

Really what I wanted is to have `::nix::fs` and `::fs`, and C++ will simply check both in order resolve `fs::foo`, but that doesn't seem to be an option. I am open to other suggestions that move us towards using namespaces more (I don't just want to remove the `fs` from `symlink_exists`).

FWIW `symlink_exists` is also named to match `std::filesystem` conventions (like `status` vs `symlink_status`). I wasn't going to literally define it in `std::filesystem` of course, but I sort of think of it as "morally" a polyfill for a missing shorthand in that API.

Comment (edolstra) [reply src/libutil/file-system.cc]: I think it's confusing to have a `nix::fs` that is almost but not quite `std::filesystem`. The doxygen warning is a symptom of that confusion (is `nix::fs::path` the same as `std::filesystem::path`?).

But I've removed the `nix::fs` changes from this PR.

**Review by Ericson2314 — APPROVED** [review]

**Pull request merged**